h5py/tests/test_dataset.py

```diff
@@ -704,6 +704,46 @@ def test_track_times(self):
         similar = self.f.create_dataset_like('lenovo', orig)
         self.assertEqual(0, h5py.h5g.get_objinfo(similar._id).mtime)
 
+class TestChunkIterator(BaseDataset):
+    def test_no_chunks(self):
+        dset = self.f.create_dataset("foo", ())
+        with self.assertRaises(TypeError):
+            dset.iter_chunks()
+
+    def test_1d(self):
+        dset = self.f.create_dataset("foo", (100,), chunks=(32,))
+        chunk_count = 0
+        expected = (slice(0,32,1), slice(32,64,1), slice(64,96,1),slice(96,100,1))
+        for s in dset.iter_chunks():
+            self.assertEqual(s, expected[chunk_count])
+            chunk_count += 1
+        self.assertEqual(chunk_count, 4)
+        chunk_count = 0
+        expected = (slice(50,64,1), slice(64,96,1), slice(96,97,1))
+        for s in dset.iter_chunks(np.s_[50:97]):
+            self.assertEqual(s, expected[chunk_count])
+            chunk_count += 1
+        self.assertEqual(chunk_count, 3)
+
+    def test_2d(self):
+        dset = self.f.create_dataset("foo", (100,100), chunks=(32,64))
+        chunk_count = 0
+        s0_expected = (slice(0,32,1), slice(32,64,1), slice(64,96,1), slice(96,100,1))
+        s1_expected = (slice(0,64,1), slice(64,100,1))
+        for s in dset.iter_chunks():
+            self.assertEqual(len(s), 2)
+            self.assertEqual(s[0], s0_expected[chunk_count // 2])
+            self.assertEqual(s[1], s1_expected[chunk_count % 2])
+            chunk_count += 1
+        self.assertEqual(chunk_count, 8)
+
+        chunk_count = 0
+        for s in dset.iter_chunks(np.s_[48:52,40:50]):
+            self.assertEqual(len(s), 2)
+            self.assertEqual(s[0], slice(48,52,1))
+            self.assertEqual(s[1], slice(40,50,1))
+            chunk_count += 1
+        self.assertEqual(chunk_count, 1)
```

Comment:
I think tests like this can be written more neatly by converting the iterator to a list, something like this:

```python
assert list(dset.iter_chunks(np.s_[48:52, 40:50])) == [
  (slice(48, 52, 1), slice(40, 50, 1))
]
```

pytest should give useful information if an assertion like this fails.

Reply:
updated.